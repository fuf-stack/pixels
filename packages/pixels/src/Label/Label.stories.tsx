import type { Meta, StoryObj } from '@storybook/react';
import type { LabelProps } from './Label';

import { FaRocket } from 'react-icons/fa6';

import { action } from '@storybook/addon-actions';

import Label from './Label';

const meta: Meta<typeof Label> = {
  title: 'pixels/Label',
  component: Label,
};

export default meta;
type Story = StoryObj<typeof Label>;

export const Default: Story = {
  args: {
    children: 'Some Label',
  },
};

export const IconOnly: Story = {
  args: {
    icon: <FaRocket />,
  },
};

export const IconAndContent: Story = {
  args: {
    icon: <FaRocket />,
    children: 'Some Label',
  },
};

export const EndContent: Story = {
  args: {
    children: 'Some Label',
    endContent: <FaRocket />,
  },
};

export const WithCloseButton: Story = {
  args: {
    children: 'Some Label',
    onClose: action('onClose'),
  },
};

export const AllColors: Story = {
  argTypes: {
    variant: {
      control: 'select',
      options: ['solid', 'bordered', 'light', 'flat', 'faded', 'dot'],
    },
  },
  render: (args) => (
    <>
      {[
        'default',
        'primary',
        'secondary',
        'info',
        'success',
        'warning',
        'danger',
      ].map((color) => (
        <div key={color} style={{ marginTop: '10px' }}>
          <Label {...args} color={color as LabelProps['color']}>
            {color}
          </Label>
        </div>
      ))}
    </>
  ),
};

export const AllSizes: Story = {
  render: () => (
    <>
      {['sm', 'md', 'lg'].map((size) => (
        <div key={size} style={{ marginTop: '10px' }}>
          <Label size={size as LabelProps['size']}>{size}</Label>
        </div>
      ))}
    </>
  ),
};

export const AllRadius: Story = {
  render: () => (
    <>
      {['none', 'sm', 'md', 'lg', 'full'].map((radius) => (
        <div key={radius} style={{ marginTop: '10px' }}>
          <Label radius={radius as LabelProps['radius']}>{radius}</Label>
        </div>
      ))}
    </>
  ),
};

export const AllVariants: Story = {
  render: () => (
    <>
      {['solid', 'bordered', 'light', 'flat', 'faded', 'dot'].map((variant) => (
        <div key={variant} style={{ marginTop: '10px' }}>
          <Label color="success" variant={variant as LabelProps['variant']}>
            {variant}
          </Label>
        </div>
      ))}
    </>
  ),
};
