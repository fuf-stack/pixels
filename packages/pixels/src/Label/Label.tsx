import type { TVClassName, TVProps } from '@fuf-stack/pixel-utils';
import type { ChipProps as HeroLabelProps } from '@heroui/chip';
import type { ReactNode } from 'react';

import { Chip as HeroLabel } from '@heroui/chip';
import { chip as heroLabelVariants } from '@heroui/theme';

import { tv, variantsToClassNames } from '@fuf-stack/pixel-utils';

// label variants
export const labelVariants = tv({
  slots: {
    base: '',
    closeButton: '',
    content: '',
    dot: '',
    icon: '',
  },
  variants: {
    color: {
      // see: https://github.com/heroui-inc/heroui/blob/canary/packages/core/theme/src/components/chip.ts
      info: {
        dot: 'bg-info',
      },
      ...heroLabelVariants.variants.color,
    },
    isIconOnly: {
      true: {
        content: 'px-1',
      },
      false: {
        content: 'flex items-center gap-2',
      },
    },
    hasEndContent: {
      true: {
        base: 'pr-2',
      },
    },
    variant: heroLabelVariants.variants.variant,
  },
  defaultVariants: heroLabelVariants.defaultVariants,
  compoundVariants: [
    // see: https://github.com/heroui-inc/heroui/blob/canary/packages/core/theme/src/components/chip.ts
    ...heroLabelVariants.compoundVariants,
    // solid / color info
    {
      variant: 'solid',
      color: 'info',
      class: {
        base: 'bg-info text-info-foreground',
      },
    },
    // bordered / color info
    {
      variant: 'bordered',
      color: 'info',
      class: {
        base: 'border-info text-info',
      },
    },
    // light / color info
    {
      variant: 'light',
      color: 'info',
      class: {
        base: 'text-info',
      },
    },
    // flat / color info
    {
      variant: 'flat',
      color: 'info',
      class: {
        base: 'bg-info/20 text-info-600',
      },
    },
    // faded / color info
    {
      variant: 'faded',
      color: 'info',
      class: {
        base: 'border-default text-info',
      },
    },
  ],
});

type VariantProps = TVProps<typeof labelVariants>;
type ClassName = TVClassName<typeof labelVariants>;

export interface LabelProps extends VariantProps {
  /** content of the label */
  children?: ReactNode;
  /** CSS class name */
  className?: ClassName;
  /** color of the label */
  color?: VariantProps['color'];
  /** element to be rendered in the right side of the label */
  endContent?: HeroLabelProps['endContent'];
  /** optional label icon, when only icon provided without children and endContent  */
  icon?: ReactNode;
  /** add close button to endContent */
  onClose?: HeroLabelProps['onClose'];
  /** radius of the label */
  radius?: HeroLabelProps['radius'];
  /** size of the label */
  size?: HeroLabelProps['size'];
  /** style variant of the label */
  variant?: 'solid' | 'bordered' | 'light' | 'flat' | 'faded' | 'dot';
}

/**
 * Label component based on [HeroUI Chip](https://www.heroui.com//docs/components/chip)
 */
const Label = ({
  children = undefined,
  className: _className = undefined,
  color = 'default',
  endContent = undefined,
  icon = undefined,
  onClose = undefined,
  radius = 'full',
  size = 'md',
  variant = 'solid',
}: LabelProps) => {
  // determine variants based on props
  const isIconOnly = !!icon && !children && !endContent;
  const hasEndContent = !!endContent;

  // classNames from slots
  const variants = labelVariants({ color, isIconOnly, hasEndContent, variant });
  const classNames = variantsToClassNames(variants, _className, 'base');

  return (
    <HeroLabel
      classNames={classNames}
      endContent={endContent}
      onClose={onClose}
      radius={radius}
      size={size}
      variant={variant}
    >
      {icon ? <span className={classNames.icon}>{icon}</span> : null}
      {children}
    </HeroLabel>
  );
};

export default Label;
