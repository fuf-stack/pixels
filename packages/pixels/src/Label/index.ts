import Label, { labelVariants } from './Label';

export type { LabelProps } from './Label';

export { Label, labelVariants };

export default Label;
