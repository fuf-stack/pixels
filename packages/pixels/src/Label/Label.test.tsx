import { describe } from 'vitest';

import storySnapshots from '@repo/storybook-config/story-snapshots';

import * as stories from './Label.stories';

describe('Story Snapshots', () => {
  storySnapshots(stories);
});
