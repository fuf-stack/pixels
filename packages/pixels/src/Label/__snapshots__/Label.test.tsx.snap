// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Story Snapshots > AllColors 1`] = `
<div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        default
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-primary text-primary-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        primary
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-secondary text-secondary-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        secondary
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-success text-success-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        success
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-warning text-warning-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        warning
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-danger text-danger-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        danger
      </span>
    </div>
  </div>
</div>
`;

exports[`Story Snapshots > AllRadius 1`] = `
<div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-none bg-default text-default-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        none
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-small bg-default text-default-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        sm
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-medium bg-default text-default-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        md
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-large bg-default text-default-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        lg
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        full
      </span>
    </div>
  </div>
</div>
`;

exports[`Story Snapshots > AllSizes 1`] = `
<div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-6 text-tiny rounded-full bg-default text-default-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-1"
      >
        sm
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        md
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-2 h-8 text-medium rounded-full bg-default text-default-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        lg
      </span>
    </div>
  </div>
</div>
`;

exports[`Story Snapshots > AllVariants 1`] = `
<div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-success text-success-foreground"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        solid
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap border-medium px-1 h-7 text-small rounded-full bg-transparent border-success text-success"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        bordered
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-transparent text-success"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        light
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-success/20 text-success-600 dark:text-success"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        flat
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap border-medium px-1 h-7 text-small rounded-full border-default bg-default-100 text-success"
    >
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        faded
      </span>
    </div>
  </div>
  <div
    style="margin-top: 10px;"
  >
    <div
      class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap border-medium border-default text-foreground bg-transparent px-1 h-7 text-small rounded-full"
    >
      <span
        class="w-2 h-2 ml-1 rounded-full bg-success"
      />
      <span
        class="flex-1 text-inherit font-normal px-2"
      >
        dot
      </span>
    </div>
  </div>
</div>
`;

exports[`Story Snapshots > Default 1`] = `
<div>
  <div
    class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground"
  >
    <span
      class="flex-1 text-inherit font-normal px-2"
    >
      Some Label
    </span>
  </div>
</div>
`;

exports[`Story Snapshots > EndContent 1`] = `
<div>
  <div
    class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground"
  >
    <span
      class="flex-1 text-inherit font-normal px-2 pr-1"
    >
      Some Label
    </span>
    <svg
      class="max-h-[80%]"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M156.6 384.9L125.7 354c-8.5-8.5-11.5-20.8-7.7-32.2c3-8.9 7-20.5 11.8-33.8L24 288c-8.6 0-16.6-4.6-20.9-12.1s-4.2-16.7 .2-24.1l52.5-88.5c13-21.9 36.5-35.3 61.9-35.3l82.3 0c2.4-4 4.8-7.7 7.2-11.3C289.1-4.1 411.1-8.1 483.9 5.3c11.6 2.1 20.6 11.2 22.8 22.8c13.4 72.9 9.3 194.8-111.4 276.7c-3.5 2.4-7.3 4.8-11.3 7.2v82.3c0 25.4-13.4 49-35.3 61.9l-88.5 52.5c-7.4 4.4-16.6 4.5-24.1 .2s-12.1-12.2-12.1-20.9V380.8c-14.1 4.9-26.4 8.9-35.7 11.9c-11.2 3.6-23.4 .5-31.8-7.8zM384 168a40 40 0 1 0 0-80 40 40 0 1 0 0 80z"
      />
    </svg>
  </div>
</div>
`;

exports[`Story Snapshots > StartContent 1`] = `
<div>
  <div
    class="relative max-w-fit min-w-min inline-flex items-center justify-between box-border whitespace-nowrap px-1 h-7 text-small rounded-full bg-default text-default-foreground"
  >
    <svg
      class="max-h-[80%]"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M156.6 384.9L125.7 354c-8.5-8.5-11.5-20.8-7.7-32.2c3-8.9 7-20.5 11.8-33.8L24 288c-8.6 0-16.6-4.6-20.9-12.1s-4.2-16.7 .2-24.1l52.5-88.5c13-21.9 36.5-35.3 61.9-35.3l82.3 0c2.4-4 4.8-7.7 7.2-11.3C289.1-4.1 411.1-8.1 483.9 5.3c11.6 2.1 20.6 11.2 22.8 22.8c13.4 72.9 9.3 194.8-111.4 276.7c-3.5 2.4-7.3 4.8-11.3 7.2v82.3c0 25.4-13.4 49-35.3 61.9l-88.5 52.5c-7.4 4.4-16.6 4.5-24.1 .2s-12.1-12.2-12.1-20.9V380.8c-14.1 4.9-26.4 8.9-35.7 11.9c-11.2 3.6-23.4 .5-31.8-7.8zM384 168a40 40 0 1 0 0-80 40 40 0 1 0 0 80z"
      />
    </svg>
    <span
      class="flex-1 text-inherit font-normal px-2 pl-1"
    >
      Some Label
    </span>
  </div>
</div>
`;
